### PR TITLE
Robot shooter

### PR DESCRIPTION
**Pull request by avidraccoon**

### CODE REVIEW

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -19,6 +21,17 @@ public class RobotContainer {
...
+            new ShooterIntakeSubsystem(
+                    new ShooterIntakeIO() {
+
+                        @Override
+                        public void setVoltage(double volts) {
+                            // TODO Auto-generated method stub
+                            throw new UnsupportedOperationException(
+                                    "Unimplemented method 'setVoltage'");
+                        }
+                    });
```
Comment (jkleiber) [inline src/main/java/frc/robot/RobotContainer.java]: We should be implementing the IO classes in separate files - one for hardware and one for sim. They both should use the interface you've created.

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -68,7 +79,14 @@ public void robotPeriodic() {
...
         // and running subsystem periodic() methods.  This must be called from the robot's periodic
         // block in order for anything in the Command-based framework to work.
+
         CommandScheduler.getInstance().run();
+        Logger.recordOutput("randomCount", pose2d);
+        Transform2d modifier = transform2d.times(Math.sin(count / 100.0));
+        pose2d = pose2d.transformBy(modifier);
+        Logger.recordOutput("joystickExpected", Math.sin(count / 100.0));
+
+        count++;
```
Comment (jkleiber) [inline src/main/java/frc/robot/Robot.java]: Is this just test code?

File: src/main/java/frc/robot/Robot.java [hunk 3]
```diff
@@ -24,6 +28,11 @@ public class Robot extends LoggedRobot {
 
     private RobotContainer robotContainer;
 
+    Pose2d pose2d = new Pose2d(0.0, 0.0, new Rotation2d(0.0, 0.0));
+    Transform2d transform2d =
+            new Transform2d(new Translation2d(0.02, 0.0), new Rotation2d(0.0, 0.0));
+    int count = 0;
+
```
Comment (jkleiber) [inline src/main/java/frc/robot/Robot.java]: Do you need pose for the shooter?

File: src/main/java/frc/robot/Robot.java [hunk 4]
```diff
@@ -52,6 +61,8 @@ public void robotInit() {
         }
 
         Logger.start();
+        Logger.recordOutput("randomTesting", new Pose2d(0.0, 0.0, new Rotation2d(0.0, 0.0)));
```
Comment (jkleiber) [inline src/main/java/frc/robot/Robot.java]: can we remove this test code before we merge this?

File: src/main/java/frc/robot/subsystems/shooter_intake/ShooterIntakeSubsystem.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+        currentState.periodic();
+    }
+
+    public void setTargetState(State target) {
+        this.targetState = target;
+    }
+
+    // spotless:off
+    public enum State {
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/shooter_intake/ShooterIntakeSubsystem.java:32]: After looking at this more, I'm not sure using the enum functions really is buying us much. It's nice that we get some inheritance here, but it's also sort of confusing to read with `changeStates` and `changeState` having two different outcomes.

One thing that I think is unfortunate is that we have to have a static instance of the subsystem inside the enum. This is not intuitive and I don't think we want to design things this way.

I think we should go back to the system we did last year for now - an enum that defines state and a switch statement that dispatches specific functions which each run the logic for their respective state.

**Review by jkleiber — COMMENTED** [review]
First pass review of the shooter

File: src/main/java/frc/robot/Robot.java (hunk 3 above)
Comment (avidraccoon) [reply src/main/java/frc/robot/Robot.java]: No sorry

File: src/main/java/frc/robot/Robot.java (hunk 4 above)
Comment (avidraccoon) [reply src/main/java/frc/robot/Robot.java]: Yes

File: src/main/java/frc/robot/Robot.java (hunk 2 above)
Comment (avidraccoon) [reply src/main/java/frc/robot/Robot.java]: Yes let me remove it quickly.

File: src/main/java/frc/robot/RobotContainer.java (hunk 1 above)
Comment (avidraccoon) [reply src/main/java/frc/robot/RobotContainer.java]: Forgot to push new commit

File: src/main/java/frc/robot/subsystems/shooter_intake/ShooterIntakeSubsystem.java (hunk 5 above)
Comment (avidraccoon) [reply src/main/java/frc/robot/subsystems/shooter_intake/ShooterIntakeSubsystem.java:32]: I believe I can fix both of those issue very quickly if there is not other issues is it fine to keep this implementation with a bit of cleaning up.

File: src/main/java/frc/robot/RobotContainer.java (hunk 1 above)
Comment (avidraccoon) [reply src/main/java/frc/robot/RobotContainer.java]: I just forgot to remove those lines of code I believe.

**Comment by avidraccoon** [comment]
I'm having problems with solving the conflicting files can you help me?

**Comment by jkleiber** [comment]
@avidraccoon what I do when I see merge conflicts is try to preserve the code that is in conflict with my code while still making sure my feature works. In this case it probably makes sense to "Accept Both" changes when doing the rebase and then refactor the code as needed to make things work.

The rebase process can be found here: https://team401.github.io/Docs/Programming/Software-Workflow.html#keep-branch-updated

If there's a specific problem I can help with just let me know!

File: src/main/java/frc/robot/subsystems/shooter_intake/ShooterIntakeSubsystem.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import org.littletonrobotics.junction.Logger;
+
+public class ShooterIntakeSubsystem extends SubsystemBase {
+
+    // Move to constants later
+    double intakeTargetRPM = -100.0;
+    double shootingTargetRPM = 300.0;
+    //
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/shooter_intake/ShooterIntakeSubsystem.java]: Let's move these to constants

We can store constants in src/main/java/frc/robot/constants and eventually JSONify everything

File: src/main/java/frc/robot/subsystems/shooter_intake/ShooterIntakeIOSim.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.wpilibj.simulation.FlywheelSim;
+import org.littletonrobotics.junction.Logger;
+
+public class ShooterIntakeIOSim implements ShooterIntakeIO {
+
+    FlywheelSim flywheelSim = new FlywheelSim(DCMotor.getCIM(2), 1.0, 1.0);
+    PIDController controller = new PIDController(0.75, 0.1, 0.0);
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/shooter_intake/ShooterIntakeIOSim.java]: If these are going to be the control gains, we should move them to constants as well

**Review by jkleiber — COMMENTED** [review]
Second pass, this looks pretty clean now. We just need to get the merge conflicts resolved and then it'll be ready to merge.

After we merge this, we need to get the hardware IO made so we can test on the robot. Can you create a ticket to do this if you haven't already?

**Comment by avidraccoon** [comment]
I moved the constants and prepared for merge but now it can't build because it can find classes for import, how do I fix this?

Edit: It can build but simulation will not run.

Edit 2: Fixed it just needed to clear the build cache

**Review by jkleiber — APPROVED** [review]
There's some refactoring to do here, but let's do that in a separate issue. Approved - please merge when you are able

**Pull request merged**